File: src/extractors/zig.ts
import type {
  Call,
  ExtractorOutput,
  SubDeclaration,
  TreeSitterNode,
  TreeSitterTree,
} from '../types.js';
import { findChild, nodeEndLine } from './helpers.js';

/**
 * Extract symbols from Zig files.
 *
 * Zig's structs/enums/unions are anonymous — their names come from the
 * enclosing `variable_declaration` (e.g. `const Foo = struct { ... };`).
 */
export function extractZigSymbols(tree: TreeSitterTree, _filePath: string): ExtractorOutput {
  const ctx: ExtractorOutput = {
    definitions: [],
    calls: [],
    imports: [],
    classes: [],
    exports: [],
    typeMap: new Map(),
  };

  walkZigNode(tree.rootNode, ctx);
  return ctx;
}

function walkZigNode(node: TreeSitterNode, ctx: ExtractorOutput): void {
  switch (node.type) {
    case 'function_declaration':
      handleZigFunction(node, ctx);
      break;
    case 'variable_declaration':
      handleZigVariable(node, ctx);
      break;
    case 'call_expression':
      handleZigCallExpression(node, ctx);
      break;
    case 'builtin_function':
      handleZigBuiltin(node, ctx);
      break;
    case 'test_declaration':
      handleZigTest(node, ctx);
      break;
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) walkZigNode(child, ctx);
  }
}

function isInsideZigContainer(node: TreeSitterNode): boolean {
  let current = node.parent;
  while (current) {
    if (current.type === 'struct_declaration' || current.type === 'union_declaration') return true;
    current = current.parent;
  }
  return false;
}

function handleZigFunction(node: TreeSitterNode, ctx: ExtractorOutput): void {
  if (isInsideZigContainer(node)) return; // already emitted by extractZigContainerMethods

  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  const params = extractZigParams(node);

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'function',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
    visibility: isZigPub(node) ? 'public' : 'private',
  });
}

function extractZigParams(funcNode: TreeSitterNode): SubDeclaration[] {
  const params: SubDeclaration[] = [];
  const paramList = funcNode.childForFieldName('parameters');
  if (!paramList) return params;

  for (let i = 0; i < paramList.childCount; i++) {
    const param = paramList.child(i);
    if (!param || param.type !== 'parameter') continue;
    const nameNode = findChild(param, 'identifier');
    if (nameNode) {
      params.push({ name: nameNode.text, kind: 'parameter', line: param.startPosition.row + 1 });
    }
  }
  return params;
}

function handleZigVariable(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = findChild(node, 'identifier');
  if (!nameNode) return;
  const name = nameNode.text;

  if (tryHandleZigContainerDecl(node, name, ctx)) return;
  if (tryHandleZigImportDecl(node, name, ctx)) return;

  // Regular constant/variable
  const isConst = hasChildText(node, 'const');
  ctx.definitions.push({
    name,
    kind: isConst ? 'constant' : 'variable',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

/** Try to handle a variable_declaration as a struct/enum/union. Returns true if handled. */
function tryHandleZigContainerDecl(
  node: TreeSitterNode,
  name: string,
  ctx: ExtractorOutput,
): boolean {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (!child) continue;
    const containerKind = zigContainerKind(child.type);
    if (!containerKind) continue;

    const members = child.type === 'struct_declaration' ? extractZigContainerFields(child) : [];
    ctx.definitions.push({
      name,
      kind: containerKind,
      line: node.startPosition.row + 1,
      endLine: nodeEndLine(node),
      children: members.length > 0 ? members : undefined,
      visibility: isZigPub(node) ? 'public' : undefined,
    });
    if (child.type === 'struct_declaration') extractZigContainerMethods(child, name, ctx);
    return true;
  }
  return false;
}

/** Map a Zig container node type to a definition kind, or null if not a container. */
function zigContainerKind(nodeType: string): 'struct' | 'enum' | undefined {
  if (nodeType === 'struct_declaration' || nodeType === 'union_declaration') return 'struct';
  if (nodeType === 'enum_declaration') return 'enum';
  return undefined;
}

/** Try to handle a variable_declaration as an @import. Returns true if handled. */
function tryHandleZigImportDecl(node: TreeSitterNode, name: string, ctx: ExtractorOutput): boolean {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (!child || child.type !== 'builtin_function') continue;
    const source = extractZigImportSource(child);
    if (source) {
      ctx.imports.push({ source, names: [name], line: node.startPosition.row + 1 });
      return true;
    }
  }
  return false;
}

/** Extract the import source path from a builtin_function node if it is @import. */
function extractZigImportSource(builtinNode: TreeSitterNode): string | null {
  const builtinId = findChild(builtinNode, 'builtin_identifier');
  if (builtinId?.text !== '@import') return null;
  const args = findChild(builtinNode, 'arguments');
  if (!args) return null;
  const strArg = findChild(args, 'string_literal') || findChild(args, 'string');
  return strArg ? strArg.text.replace(/^"|"$/g, '') : null;
}

function extractZigContainerFields(container: TreeSitterNode): SubDeclaration[] {
  const fields: SubDeclaration[] = [];
  for (let i = 0; i < container.childCount; i++) {
    const child = container.child(i);
    if (!child || child.type !== 'container_field') continue;
    const nameNode = child.childForFieldName('name') || findChild(child, 'identifier');
    if (nameNode) {
      fields.push({ name: nameNode.text, kind: 'property', line: child.startPosition.row + 1 });
    }
  }
  return fields;
}

function extractZigContainerMethods(
  container: TreeSitterNode,
  parentName: string,
  ctx: ExtractorOutput,
): void {
  for (let i = 0; i < container.childCount; i++) {
    const child = container.child(i);
    if (!child || child.type !== 'function_declaration') continue;
    const nameNode = child.childForFieldName('name');
    if (nameNode) {
      ctx.definitions.push({
        name: `${parentName}.${nameNode.text}`,
        kind: 'method',
        line: child.startPosition.row + 1,
        endLine: nodeEndLine(child),
        visibility: isZigPub(child) ? 'public' : 'private',
      });
    }
  }
}

function handleZigCallExpression(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const funcNode = node.childForFieldName('function');
  if (!funcNode) return;

  const call: Call = { name: '', line: node.startPosition.row + 1 };

  if (funcNode.type === 'field_expression' || funcNode.type === 'field_access') {
    const field = funcNode.childForFieldName('field') || funcNode.childForFieldName('member');
    const value = funcNode.childForFieldName('value') || funcNode.child(0);
    if (field) call.name = field.text;
    if (value) call.receiver = value.text;
  } else {
    call.name = funcNode.text;
  }

  if (call.name) ctx.calls.push(call);
}

function handleZigBuiltin(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const builtinId = findChild(node, 'builtin_identifier');
  if (!builtinId) return;

  // Treat @import as import (when standalone, not in variable_declaration)
  if (builtinId.text === '@import' && node.parent?.type !== 'variable_declaration') {
    const args = findChild(node, 'arguments');
    if (args) {
      const strArg = findChild(args, 'string_literal') || findChild(args, 'string');
      if (strArg) {
        const source = strArg.text.replace(/^"|"$/g, '');
        ctx.imports.push({
          source,
          names: ['@import'],
          line: node.startPosition.row + 1,
        });
      }
    }
    return;
  }

  // Other builtins are calls
  ctx.calls.push({ name: builtinId.text, line: node.startPosition.row + 1 });
}

function handleZigTest(node: TreeSitterNode, ctx: ExtractorOutput): void {
  let name = 'test';
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (!child) continue;
    if (child.type === 'string_literal' || child.type === 'string') {
      // Extract the string content child if available, otherwise strip quotes
      const content = findChild(child, 'string_content');
      name = content ? content.text : child.text.replace(/^"|"$/g, '');
      break;
    }
    if (child.type === 'identifier') {
      name = child.text;
      break;
    }
  }

  ctx.definitions.push({
    name,
    kind: 'function',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function isZigPub(node: TreeSitterNode): boolean {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child && child.type === 'pub') return true;
    if (child && child.text === 'pub') return true;
  }
  return false;
}

function hasChildText(node: TreeSitterNode, text: string): boolean {
  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child && child.text === text) return true;
  }
  return false;
}

File: crates/codegraph-core/src/extractors/zig.rs
use tree_sitter::{Node, Tree};
use crate::ast_analysis::cfg::build_function_cfg;
use crate::ast_analysis::complexity::compute_all_metrics;
use crate::types::*;
use super::helpers::*;
use super::SymbolExtractor;

pub struct ZigExtractor;

impl SymbolExtractor for ZigExtractor {
    fn extract(&self, tree: &Tree, source: &[u8], file_path: &str) -> FileSymbols {
        let mut symbols = FileSymbols::new(file_path.to_string());
        walk_tree(&tree.root_node(), source, &mut symbols, match_zig_node);
        walk_ast_nodes_with_config(&tree.root_node(), source, &mut symbols.ast_nodes, &ZIG_AST_CONFIG);
        symbols
    }
}

fn match_zig_node(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "function_declaration" => handle_zig_function(node, source, symbols),
        "variable_declaration" => handle_zig_variable(node, source, symbols),
        "call_expression" => handle_zig_call(node, source, symbols),
        "builtin_function" => handle_zig_builtin(node, source, symbols),
        "test_declaration" => handle_zig_test(node, source, symbols),
        _ => {}
    }
}

fn handle_zig_function(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    let parent_struct = find_zig_parent_struct(node, source);
    let name_text = node_text(&name_node, source);
    let (full_name, kind) = match &parent_struct {
        Some(s) => (format!("{}.{}", s, name_text), "method"),
        None => (name_text.to_string(), "function"),
    };

    let params = extract_zig_params(node, source);
    symbols.definitions.push(Definition {
        name: full_name,
        kind: kind.to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: compute_all_metrics(node, source, "zig"),
        cfg: build_function_cfg(node, "zig", source),
        children: opt_children(params),
    });
}

fn extract_zig_params(func_node: &Node, source: &[u8]) -> Vec<Definition> {
    let mut params = Vec::new();
    if let Some(param_list) = func_node.child_by_field_name("parameters") {
        for i in 0..param_list.child_count() {
            if let Some(child) = param_list.child(i) {
                if child.kind() == "parameter" {
                    if let Some(name_node) = find_child(&child, "identifier") {
                        params.push(child_def(
                            node_text(&name_node, source).to_string(),
                            "parameter",
                            start_line(&child),
                        ));
                    }
                }
            }
        }
    }
    params
}

fn handle_zig_variable(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let Some(name_node) = find_child(node, "identifier") else { return };
    let name = node_text(&name_node, source).to_string();

    // Check for struct/enum/union type definition
    if try_handle_zig_type_def(node, source, symbols, &name) {
        return;
    }

    // Check for @import binding
    if try_handle_zig_import(node, source, symbols, name.clone()) {
        return;
    }

    // Regular const/var
    let is_const = node_has_child_text(node, source, "const");
    symbols.definitions.push(Definition {
        name,
        kind: if is_const { "constant" } else { "variable" }.to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn try_handle_zig_type_def(node: &Node, source: &[u8], symbols: &mut FileSymbols, name: &str) -> bool {
    for i in 0..node.child_count() {
        let Some(child) = node.child(i) else { continue };
        let (kind, children) = match child.kind() {
            "struct_declaration" => ("struct", opt_children(extract_zig_container_fields(&child, source))),
            "enum_declaration" => ("enum", None),
            "union_declaration" => ("struct", None),
            _ => continue,
        };
        symbols.definitions.push(Definition {
            name: name.to_string(),
            kind: kind.to_string(),
            line: start_line(node),
            end_line: Some(end_line(node)),
            decorators: None,
            complexity: None,
            cfg: None,
            children,
        });
        return true;
    }
    false
}

fn try_handle_zig_import(node: &Node, source: &[u8], symbols: &mut FileSymbols, name: String) -> bool {
    for i in 0..node.child_count() {
        let Some(child) = node.child(i) else { continue };
        if child.kind() != "builtin_function" { continue; }
        if let Some(path) = extract_zig_import_path(&child, source) {
            symbols.imports.push(Import::new(path, vec![name], start_line(node)));
            return true;
        }
    }
    false
}

fn extract_zig_import_path(builtin: &Node, source: &[u8]) -> Option<String> {
    let builtin_id = find_child(builtin, "builtin_identifier")?;
    if node_text(&builtin_id, source) != "@import" { return None; }
    let args = find_child(builtin, "arguments")?;
    for j in 0..args.child_count() {
        let Some(arg) = args.child(j) else { continue };
        if arg.kind() == "string_literal" || arg.kind() == "string" {
            let raw = node_text(&arg, source);
            return Some(raw.trim_matches('"').to_string());
        }
    }
    None
}

fn extract_zig_container_fields(container: &Node, source: &[u8]) -> Vec<Definition> {
    let mut fields = Vec::new();
    for i in 0..container.child_count() {
        if let Some(child) = container.child(i) {
            if child.kind() == "container_field" {
                let name_node = child.child_by_field_name("name")
                    .or_else(|| find_child(&child, "identifier"));
                if let Some(n) = name_node {
                    fields.push(child_def(
                        node_text(&n, source).to_string(),
                        "property",
                        start_line(&child),
                    ));
                }
            }
        }
    }
    fields
}

fn handle_zig_call(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let func_node = match node.child_by_field_name("function").or_else(|| node.child(0)) {
        Some(n) => n,
        None => return,
    };

    match func_node.kind() {
        "field_expression" | "field_access" => {
            let field = func_node.child_by_field_name("field")
                .or_else(|| func_node.child_by_field_name("member"));
            let value = func_node.child(0);
            if let Some(f) = field {
                symbols.calls.push(Call {
                    name: node_text(&f, source).to_string(),
                    line: start_line(node),
                    dynamic: None,
                    receiver: value.map(|v| node_text(&v, source).to_string()),
                    ..Default::default()
                });
            }
        }
        _ => {
            symbols.calls.push(Call {
                name: node_text(&func_node, source).to_string(),
                line: start_line(node),
                dynamic: None,
                receiver: None,
                ..Default::default()
            });
        }
    }
}

fn handle_zig_builtin(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let builtin_id = match find_child(node, "builtin_identifier") {
        Some(n) => n,
        None => return,
    };

    let name = node_text(&builtin_id, source);
    if name == "@import" && node.parent().map(|p| p.kind()) != Some("variable_declaration") {
        if let Some(args) = find_child(node, "arguments") {
            for i in 0..args.child_count() {
                if let Some(arg) = args.child(i) {
                    if arg.kind() == "string_literal" || arg.kind() == "string" {
                        let raw = node_text(&arg, source);
                        let source_path = raw.trim_matches('"').to_string();
                        symbols.imports.push(Import::new(
                            source_path,
                            vec!["@import".to_string()],
                            start_line(node),
                        ));
                        return;
                    }
                }
            }
        }
    }

    symbols.calls.push(Call {
        name: name.to_string(),
        line: start_line(node),
        dynamic: None,
        receiver: None,
        ..Default::default()
    });
}

fn handle_zig_test(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let mut name = "test".to_string();
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if child.kind() == "string_literal" || child.kind() == "identifier" {
                name = node_text(&child, source).trim_matches('"').to_string();
                break;
            }
        }
    }

    symbols.definitions.push(Definition {
        name,
        kind: "function".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn find_zig_parent_struct<'a>(node: &Node<'a>, source: &[u8]) -> Option<String> {
    let mut current = node.parent();
    while let Some(parent) = current {
        if parent.kind() == "struct_declaration" || parent.kind() == "union_declaration" {
            // The name is in the grandparent variable_declaration
            if let Some(gp) = parent.parent() {
                if gp.kind() == "variable_declaration" {
                    if let Some(name_node) = find_child(&gp, "identifier") {
                        return Some(node_text(&name_node, source).to_string());
                    }
                }
            }
        }
        current = parent.parent();
    }
    None
}

// TODO: wire into Definition once the struct gains a `visibility` field
#[allow(dead_code)]
fn is_zig_pub(node: &Node, source: &[u8]) -> bool {
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if node_text(&child, source) == "pub" {
                return true;
            }
        }
    }
    false
}

fn node_has_child_text(node: &Node, source: &[u8], text: &str) -> bool {
    for i in 0..node.child_count() {
        if let Some(child) = node.child(i) {
            if node_text(&child, source) == text {
                return true;
            }
        }
    }
    false
}
